### ISSUE TITLE
[Question] Why GatedTrainingSAE do use a detached forward pass to compute the auxiliary loss?

### ISSUE THREAD

**Issue by jasonrichdarmawan** [jbloomAus/SAELens#539]
# Questions

Wouldn't the goal of `aux_recon_loss` is to regularize the `pi_gate_act`? So why do we penalize `W_dec` and `b_dec`?

**Comment by chanind** [comment]
This is probably a bug, I think you're right that it makes sense to detach. it looks like dictionary_learning detaches W_dec and b_dec: https://github.com/saprmarks/dictionary_learning/blob/main/dictionary_learning/trainers/gdm.py#L80C2-L80C96

**Comment by jasonrichdarmawan** [comment]
> I think you're right that it makes sense to detach

Yes. However, when `self.W_dec.detach()` and `self.b_dec.detach()`, with the same training hyper parameters, the training failed. Either because of:
1. need training hyper parameters search
2. SAE can't learn this data distribution
3. Somewhere in `GatedTrainingSAE`, compensate the non `.detach()` auxilirary loss

Meanwhile, without `.detach()`, the training is successful

PS: I have not do autointerp yet to the non `.detach()` version

If you are interested you can check below for run variants `gated_detach`

[https://wandb.ai/jasonrichdarma/sonar_sae?nw=nwuserjasonrichdarmawan](WandB)

**Comment by chanind** [comment]
oh interesting, good you're checking! Your detach version was just modifying the following line, yeah?

```
via_gate_reconstruction = pi_gate_act @ self.W_dec + self.b_dec
```

to

```
via_gate_reconstruction = pi_gate_act @ self.W_dec.detach() + self.b_dec.detach()
```

**Comment by jasonrichdarmawan** [comment]
> oh interesting, good you're checking! Your detach version was just modifying the following line, yeah?

Yes

From what I understand, `.detach()` prevents the decoder to cheat by learning to reconstruct the input from `pi_gate_act` instead of `acts_post`. 

Is my assumption correct?

Since `W_dec` can't cheat to reconstruct input from `pi_gate_act`, there is bound for dead neurons. Unfortunately, `SAELens` library doesn't have logic for resampling weights
